src/widget/wbattery.cpp

```diff
@@ -0,0 +1,173 @@
+        }
+    }
+
+    if (context.hasNode(node, "PixmapsDischarging")) {
+        // TODO(XXX) inline SVG support via context.getPixmapSource.
+        QString dischargingPath = context.selectString(node, "PixmapsDischarging");
+        Paintable::DrawMode mode = Paintable::DrawModeFromString(
+                context.selectAttributeString(
+                        context.selectElement(node, "PixmapsDischarging"),
+                        "scalemode", "TILE"));
+        for (int i = 0; i < m_dischargingPixmaps.size(); ++i) {
+            PixmapSource source(dischargingPath.arg(i));
+            setPixmap(&m_dischargingPixmaps[i], source, mode);
+        }
+    }
+
+    if (m_pBattery) {
+        m_pBattery->update();
+    }
+}
+
+QString formatMinutes(int minutes) {
+    if (minutes < 60) {
+        return QObject::tr("%1 minutes").arg(minutes);
+    }
+    return QObject::tr("%1:%2").arg(minutes/60)
+            .arg(minutes%60, 2, 10, QChar('0'));
+}
+
+int pixmapIndexFromPercentage(double dPercentage, int numPixmaps) {
+    // See WDisplay::getActivePixmapIndex for more info on this.
+    int result = static_cast<int>(dPercentage * numPixmaps - 0.00001);
+    result = math_min(numPixmaps - 1, math_max(0, result));
+    return result;
+}
+
+void WBattery::update() {
+    int minutesLeft = m_pBattery ? m_pBattery->getMinutesLeft() : 0;
+    Battery::ChargingState chargingState = m_pBattery ?
+            m_pBattery->getChargingState() : Battery::UNKNOWN;
+    double dPercentage = m_pBattery ? m_pBattery->getPercentage() : 0;
+
+    m_pCurrentPixmap.clear();
+    switch (chargingState) {
+        case Battery::CHARGING:
+            if (!m_chargingPixmaps.isEmpty()) {
+                m_pCurrentPixmap = m_chargingPixmaps[
+                    pixmapIndexFromPercentage(dPercentage,
+                                              m_chargingPixmaps.size())];
+            }
+            if (minutesLeft == -1) {
+                setBaseTooltip(tr("Time until charged unknown."));
+            } else {
+                setBaseTooltip(tr("Time until charged: %1")
+                               .arg(formatMinutes(minutesLeft)));
+            }
+            break;
+        case Battery::DISCHARGING:
+            if (!m_dischargingPixmaps.isEmpty()) {
+                m_pCurrentPixmap = m_dischargingPixmaps[
+                    pixmapIndexFromPercentage(dPercentage,
+                                              m_dischargingPixmaps.size())];
+            }
+            if (minutesLeft == -1) {
+                setBaseTooltip(tr("Time left unknown."));
+            } else {
+                setBaseTooltip(tr("Time left: %1")
+                               .arg(formatMinutes(minutesLeft)));
+            }
+            break;
+        case Battery::CHARGED:
+            m_pCurrentPixmap = m_pPixmapCharged;
+            setBaseTooltip(tr("Battery fully charged."));
+            break;
+        case Battery::UNKNOWN:
```

Comment:
Is this state uses for desktop devices? 
We should change the Pixmap for this case as well.